runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkItemStatusClient.java

```diff
@@ -266,27 +266,62 @@ private synchronized WorkItemStatus createStatusUpdate(boolean isFinal) {
     return status;
   }
 
-  // todo(migryz) this method should return List<CounterUpdate> instead of updating member variable
   @VisibleForTesting
   synchronized void populateCounterUpdates(WorkItemStatus status) {
     if (worker == null) {
       return;
     }
 
+    // Checking against boolean, because getCompleted can return null?
     boolean isFinalUpdate = Boolean.TRUE.equals(status.getCompleted());
 
-    ImmutableList.Builder<CounterUpdate> counterUpdatesListBuilder = ImmutableList.builder();
+    Map<Object, CounterUpdate> counterUpdatesMap = new HashMap<>();
     // Output counters
-    counterUpdatesListBuilder.addAll(extractCounters(worker.getOutputCounters()));
+    extractCounters(worker.getOutputCounters())
+        .forEach(
+            x ->
+                counterUpdatesMap.put(
+                    x.getStructuredNameAndMetadata() == null
+                        ? x.getNameAndKind()
+                        : x.getStructuredNameAndMetadata(),
+                    x));
+
     // User metrics reported in Worker
-    counterUpdatesListBuilder.addAll(extractMetrics(isFinalUpdate));
+    extractMetrics(isFinalUpdate)
+        .forEach(
+            x ->
+                counterUpdatesMap.put(
+                    x.getStructuredNameAndMetadata() == null
+                        ? x.getNameAndKind()
+                        : x.getStructuredNameAndMetadata(),
+                    x));
+
+    // counterUpdatesListBuilder.addAll(extractMetrics(isFinalUpdate));
     // MSec counters reported in worker
-    counterUpdatesListBuilder.addAll(extractMsecCounters(isFinalUpdate));
-    // Metrics reported in SDK runner.
-    counterUpdatesListBuilder.addAll(worker.extractMetricUpdates());
+    extractMsecCounters(isFinalUpdate)
+        .forEach(
+            x ->
+                counterUpdatesMap.put(
+                    x.getStructuredNameAndMetadata() == null
+                        ? x.getNameAndKind()
+                        : x.getStructuredNameAndMetadata(),
+                    x));
 
-    ImmutableList<CounterUpdate> counterUpdates = counterUpdatesListBuilder.build();
```

Comment:
This is not an error in this implementation.
I overwrite all metrics that are defined within java worker with metrics from SDK intentionally.
Do you mean to log warning every time we overwrite CounterUpdate in case of coding error?

It should be transparent regardless of mode we are running in.